solr/core/src/java/org/apache/solr/cloud/ClusterSingleton.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.cloud;
+
+/**
+ * Intended for {@link org.apache.solr.core.CoreContainer} plugins that should be
+ * enabled only one instance per cluster.
+ * <p>Components that implement this interface are always in one of two states:
+ * <ul>
+ *   <li>STOPPED - the default state. The component is idle and does not perform
```

Comment:
Should we add `STARTING` and `STOPPING` states? Assuming the call to `start()` waits until the plugin has completed its startup (and similarly the call to `stop()` waiting for it to stop) might be expensive, and as implemented `start()` delays the `Overseer` starting in general.

Possibly waiting for `stop()` to complete makes sense (in order to guarantee no two `ClusterSingleton` plugins are running concurrently on the cluster, so maybe state `STOPPING` is not needed), but I'd think we don't need to wait for `start()` to have completed.

Reply:
I didn't want to complicate the API to include support for more complex state and async operations. Besides, there's only one parent that is supposed to call the `start()` method, and no other actor is supposed to do that, so from the POV of other components the singleton is either running or not, and the STARTING state is beyond their control anyway - from their POV the singleton is in "not running" state during startup.